src/API.jl

```diff
@@ -1208,13 +1208,27 @@ function precompile(ctx::Context, pkgs::Vector{PackageSpec}; internal_call::Bool
 
     # find and guard against circular deps
     circular_deps = Base.PkgId[]
-    function in_deps(_pkgs, deps, dmap)
-        isempty(deps) && return false
-        !isempty(intersect(_pkgs, deps)) && return true
-        return any(dep->in_deps(vcat(_pkgs, dep), dmap[dep], dmap), deps)
+    depths = Dict{Base.PkgId, Int}(pkg => 0 for pkg in keys(depsmap))
+    function dfs_scan!(pkg, dmap, depth=0)
+        for dep in dmap[pkg]
+            already_visited = true
+            depdepth = get!(depths, dep) do
+                already_visited = false
+                depth + 1
+            end
+            # Will be visited at top level
+            depdepth == 0 && continue
+            if depdepth < depth+1
+                # Found cycle
```

Comment:
```suggestion
                # Found cycle
                depths[dep] = typemax(Int)
```

Reply:
Actually, a state enum might be easier for this (not-visited, visiting, cycle, not-cycle) since it does not need the depth itself (as it would if this was the full Tarjan's SCC for example instead)